src/orion/algo/robo/ablr/ablr_model.py

```diff
@@ -0,0 +1,589 @@
+        """Return the predictive distribution for the given points."""
+        with torch.no_grad():
+            if self._predict_dist is None:
+                # re-create the predictive distribution.
+                # NOTE: This gets overwritten every time a new point is observed, so we know for
+                # sure that this is always the most up-to-date predictions.
+                _, self._predict_dist = self(self.X, self.y)
+            y_pred_dist = self._predict_dist(x_test)
+        return y_pred_dist
+
+    def forward(
+        self, x_train: Tensor, y_train: Tensor
+    ) -> tuple[Tensor, Callable[[Tensor], Normal]]:
+        """Return the training loss and a function that will give the
+        predictive distribution over y for a given (un-normalized) input x_test.
+
+        NOTE: Assumes that `x_train`, `y_train` and an eventual `x_test` are
+        NOT already rescaled/normalized!
+        """
+        x_train = torch.as_tensor(x_train)
+        y_train = torch.as_tensor(y_train)
+        # TODO: Do we normalize the portion that has the contextual information?
+        if all(self.x_mean == 0):
+            self.x_mean = x_train.mean(0)
+        if all(self.x_var == 1):
+            x_var = x_train.var(0)
+            # minimum variance, because the context vectors are actually always
+            # the same, so the variance along that dimension is 0, which makes
+            # NaN values in the normalized inputs.
+            self.x_var = torch.maximum(x_var, torch.zeros_like(x_var) + 1e-5)
+        if self.y_mean == 0:
+            self.y_mean = y_train.mean(0)
+        if self.y_var == 1:
+            y_var = y_train.var(0)
+            self.y_var = torch.maximum(y_var, torch.zeros_like(y_var) + 1e-5)
+
+        assert x_train.shape[-1] == self.x_mean.shape[-1], (x_train.shape, self.x_mean)
+        x_train = torch.as_tensor(x_train, dtype=torch.float)
+        y_train = torch.as_tensor(y_train, dtype=torch.float)
+        x_train = self._normalize_x(x_train) if self.normalize_inputs else x_train
+        y_train = self._normalize_y(y_train) if self.normalize_inputs else y_train
+        neg_mll, predictive_distribution_fn = self.predictive_mean_and_variance(
+            x_train, y_train
+        )
+
+        def predict_normalized(x_test: Tensor | np.ndarray) -> Normal:
+            x_test = torch.as_tensor(x_test, dtype=torch.float)
+            if self.normalize_inputs:
+                x_test = self._normalize_x(x_test)
+            y_pred_dist = predictive_distribution_fn(x_test)
+            y_pred_dist = self._unnormalize_y(y_pred_dist)
+            return y_pred_dist
+
+            # y_pred = y_pred_dist.rsample()
+            # return self.unnormalize_y(y_pred)
+
+        return neg_mll, predict_normalized
+
+    def _get_loss(self, x: np.ndarray, y: np.ndarray) -> Tensor:
+        """Gets the loss for the given input and target."""
+        return self(x, y)[0]
+
+    @overload
```

Comment:
This is because the class is inheriting from both nn.Module and pybnn.BaseModel? Would it be simpler to have a module object attribute in this class instead of combining both?

Reply:
Hmm. Yeah you're probably right, something like `RoBO_ABLR -> ABLR -> AblrNetwork` ?

This solution is still kinda neat though, right? :smile:

Reply:
To be honest I feel it's a bit like taking `**kwargs` as arguments for a class constructor and setting them in a loop (like we did in Trial). It looks like a great idea at first as you reduce the lines of codes to do it, but after some time the obscure signature confuses so much people that it outweighs the time saved writing these lines. I'll be happy when we fix this in Trial. :)

Reply:
Well in this case, the signature is very very well defined (with the @overload s above)!